### PR TITLE
Improve `report_errors` output

### PR DESCRIPTION
**Pull request by msimberg**
- Always print a message, even if no tests failed. This makes it easier to see that the test actually ran successfully. It also prints how many tests were run.
- Print a message and fail if `report_errors` is called but no tests were actually run.

Example outputs below. Success:
```
All tests passed. Ran 0 sanity checks and 150121 tests.
```
Failure:
```
Tests failed. 0/0 sanity checks and 1/13 tests failed.
```
No tests run:
```
No tests run. Did you forget to add tests?
```
Ideas to improve these messages are warmly welcome.

Related: #367. Not removing them yet, but I think we can quite safely remove the sanity checks (since they're unused).

Made a few tests that were triggering the "No tests run" case into compile-only tests. Added a few `PIKA_TEST(true)`s to others.

### CODE REVIEW

**Comment by msimberg** [comment]
bors try

**Comment by msimberg** [comment]
bors try

**Comment by msimberg** [comment]
bors try

File: testing/src/testing.cpp [hunk 1]
```diff
@@ -23,10 +23,28 @@
...
+            case counter_sanity:
+                ++sanity_tests_;
+                return;
+            case counter_test:
+                ++test_tests_;
+                return;
+            default:
+                break;
+            }
+            PIKA_ASSERT(false);
```
Comment (aurianer) [inline testing/src/testing.cpp:44]: can the `PIKA_ASSERT(false)` be in the `default:`?

**Review by aurianer — APPROVED** [review]
LGTM, thanks a lot!

File: testing/src/testing.cpp (hunk 1 above)
Comment (msimberg) [reply testing/src/testing.cpp:44]: Yes, I think so.

**Comment by aurianer** [comment]
bors merge

File: testing/src/testing.cpp (hunk 1 above)
Comment (msimberg) [reply testing/src/testing.cpp:44]: I did not update this yet...

**Pull request merged**